// Copyright (C) 2019-2020 Siodb GmbH. All rights reserved.
// Use of this source code is governed by a license that can be found
// in the LICENSE file.

// @file Contains protobuf messages used to communicate between Siodb server
// and IO manager.

syntax = "proto3";

import "CommonTypes.proto";

package siodb.iomgr_protocol;

option optimize_for=SPEED;

/** Request message. */
message DatabaseEngineRequest {

    /** Request ID */
    uint64 request_id = 1;

    /** Request text */
    string text = 2;
}

/** Tag key-value pair. */
message Tag {
    /** Tag name */
    string name = 1;

    /** Tag value variants */
    oneof value {
        /** Boolean value */
        bool bool_value = 2;

        /** String value */
        string string_value = 3;
    }
}

/** Response from IO Manager. */
message DatabaseEngineResponse {

    /** Corresponding request ID. */
    uint64 request_id = 1;

    /** Messages from IO manager. */
    repeated StatusMessage message = 2;

    /** Free text messages without status code. */
    repeated string freetext_message = 3;

    /** Description of columns in this data set. */
    repeated ColumnDescription column_description = 4;

    /** Number of this response in series of responses. Starts with 0. */
    uint32 response_id = 5;

    /**
     * Number of reponses in series.
     * May be sent only in first response.
     * Zero value means single response.
     */
    uint32 response_count = 6;

    /** Indicates that affected row count is present. */
    bool has_affected_row_count = 7;

    /** Affected row count. */
    uint64 affected_row_count = 8;

    /** Tags. */
    repeated Tag tag = 9;
}

/** Begin authentication request */
message BeginAuthenticateUserRequest {

    /** User request */
    string user_name = 1;
}

/** Begin authentication response */
message BeginAuthenticateUserResponse {

    /** Indication that session is started */
    bool session_started = 1;

    /** Message from IO manager. Set in case of error. */
    StatusMessage message = 2;
}

/** Authenticate user request */
message AuthenticateUserRequest {

    /** Challenge */
    bytes challenge = 1;

    /** Signed challenge user from a user */
    bytes signature = 2;
}

/** Authenticate user Response */
message AuthenticateUserResponse {

    /** Indication that user is authenticated. */
    bool authenticated = 1;

    /** Message from IO manager. Set in case of error. */
    StatusMessage message = 2;

    /** ID of started session. */
    string session_id = 3;
}